### PR TITLE
[Intel GPU] Enable oneDNN GEMM Primitive

### PR DESCRIPTION
**Pull request by ZhiweiYan-96**
# Motivation 

This PR is a part of RFC #114848, and it  is a successor PR of #116249 and #116019. This PR would depend on oneDNN compilation in #116249. Some runtime support is needed in #116019.

This PR contains oneDNN GEMM support, which consists of oneDNN glue code `Matmul.h` defined `aten/src/ATen/native/mkldnn/xpu/detail` directory. Aten operators like `linear`, `baddmm` is defined in `Linear.cpp, Blass.cpp` in `aten/src/ATen/native/mkldnn/xpu/`. 

Accompanied with these files provide core functionaliy, `BlasImpl.h`, `Utils.h` and other file provide basic utilities for them. For instance, `Utils.h` provide common memory descriptor query utils for `Matmul.h` and these utility function will also be used in other primitive, like `convolution`.

In next stage, we would continually complete the oneDNN support through enabling  `matmul fusion`  and `convolution` related code.

@EikanWang  @gujinghui  @arthuryuan1987 @riverliuintel

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen

### CODE REVIEW

File: aten/src/ATen/native/mkldnn/xpu/detail/Matmul.h [hunk 1]
```diff
@@ -28,9 +28,9 @@ static inline void matmul(
       "oneDNN input matrixes must have the same ranks");
   TORCH_CHECK(result.defined(), "oneDNN matmul result should be defined");
 
-  // at::Device curDevice = at::Device(at::kXPU, current_device());
-  // auto engine = GpuEngineManager::Instance().get_engine(curDevice);
-  // auto strm = GpuStreamManager::Instance().get_stream();
+  at::Device curDevice = at::Device(at::kXPU, current_device());
```
Comment (EikanWang) [inline aten/src/ATen/native/mkldnn/xpu/detail/Matmul.h:31]: The code style is not aligned with other variables.
```suggestion
  at::Device cur_device = at::Device(at::kXPU, current_device());
```

File: aten/src/ATen/native/mkldnn/xpu/detail/Matmul.h [hunk 2]
```diff
@@ -28,9 +28,9 @@ static inline void matmul(
...
+  auto engine = GpuEngineManager::Instance().get_engine(curDevice);
+  auto strm = GpuStreamManager::Instance().get_stream();
```
Comment (EikanWang) [inline aten/src/ATen/native/mkldnn/xpu/detail/Matmul.h:33]: `strm` is not a common abbreviation. Please refine the word.

**Review by EikanWang — CHANGES_REQUESTED** [review]

**Comment by EikanWang** [comment]
Please fix the code style issue.

File: aten/src/ATen/native/mkldnn/xpu/detail/Matmul.h (hunk 2 above)
Comment (ZhiweiYan-96) [reply aten/src/ATen/native/mkldnn/xpu/detail/Matmul.h]: The `strm` has modifed to `stream`, thanks!

File: aten/src/ATen/native/mkldnn/xpu/detail/Matmul.h (hunk 1 above)
Comment (ZhiweiYan-96) [reply aten/src/ATen/native/mkldnn/xpu/detail/Matmul.h]: Has modified to `cur_device`, thanks.

File: aten/src/ATen/CMakeLists.txt [hunk 3]
```diff
@@ -226,6 +228,21 @@ else()
...
+  list(APPEND ATen_XPU_INCLUDE ${CMAKE_CURRENT_SOURCE_DIR}/native/mkldnn/xpu)
+  list(APPEND ATen_XPU_INCLUDE ${PROJECT_SOURCE_DIR}/third_party/ideep/mkl-dnn/include)
+  list(APPEND ATen_XPU_INCLUDE ${XPU_MKLDNN_INCLUDE})
+
+  # TODO: Remove hand-written link
+  list(APPEND ATen_XPU_INCLUDE /path/to/oneapi/include/)
+  list(APPEND ATen_XPU_INCLUDE /path/to/oneapi/include/sycl/)
+  list(APPEND ATen_XPU_DEPENDENCY_LIBS /path/to/oneapi/lib/libsycl.so.7)
+  list(APPEND ATen_XPU_DEPENDENCY_LIBS /path/to/libOpenCL.so.1)
+  list(APPEND ATen_XPU_DEPENDENCY_LIBS /path/to/libintlc.so) # Required by sycl
```
Comment (jgong5) [inline aten/src/ATen/CMakeLists.txt:243]: Please fix these "/path/to" things.

**Review by jgong5 — CHANGES_REQUESTED** [review]
I cannot find the aten op registrations on top of these added functions. Can you add them?

**Comment by EikanWang** [comment]
@ZhiweiYan-96 , since the @guangyey's first PR has landed, please rebase your code and fix the conflict.

**Comment by EikanWang** [comment]
@ZhiweiYan-96 , I suppose this PR should be closed as you have duplicated this PR and stacked this PR with other oneDNN-related PRs by ghstack.

**Comment by ZhiweiYan-96** [comment]
Thanks for reminding @EikanWang 
Close this PR, as we have duplicate a PR using ghstack here https://github.com/pytorch/pytorch/pull/117112 and https://github.com/pytorch/pytorch/pull/117202

**Pull request closed without merging**